File: medium/2616_minimize_the_maximum_difference_of_pairs/solution.cpp
#include <algorithm>
#include <vector>

class Solution {
public:
  int minimizeMax(std::vector<int> &nums, int p) {
    std::sort(nums.begin(), nums.end());

    int left = 0;
    int right = nums.back() - nums.front();

    while (left < right) {
      int mid = (left + right) / 2;

      int count = 0;
      for (size_t i = 1; i < nums.size(); ++i) {
        if (nums[i] - nums[i - 1] <= mid) {
          ++count;
          ++i;
        }
      }

      if (count >= p)
        right = mid;
      else
        left = mid + 1;
    }

    return left;
  }
};

File: medium/2616_minimize_the_maximum_difference_of_pairs/solution.py
from typing import List


class Solution:
    def minimizeMax(self, nums: List[int], p: int) -> int:
        nums.sort()

        left = 0
        right = nums[-1] - nums[0]

        while left < right:
            mid = (left + right) // 2

            count = 0
            i = 1
            while i < len(nums):
                if nums[i] - nums[i - 1] <= mid:
                    count += 1
                    i += 1

                i += 1

            if count >= p:
                right = mid
            else:
                left = mid + 1

        return left
